[Identity] Always include the Broker as part of DAC

JonathanCrd opened the pull request:
Fixes https://github.com/Azure/azure-sdk-for-net/issues/52138

This PR changes the DAC Factory to always include the Broker even if the Azure.Identity.Broker package is not installed. In order to do this, a new internal `BrokerCredential` type is used, so instead of throwing an error during the construction of DAC, an error will be thrown when `GetToken()` is called.

This is the same Behaviour as `VisualStudioCodeCredential`.

JonathanCrd committed:
Wrap the Broker in DAC in new internal type

JonathanCrd committed:
Update test

JonathanCrd committed:
Update tests

JonathanCrd committed:
Merge branch 'main' into identity-always-add-broker-in-dac

JonathanCrd committed:
Update CreateBrokerCredential instances after merge with main

JonathanCrd committed:
Update tests

christothes commented on sdk/identity/Azure.Identity/src/Credentials/BrokerCredential.cs:
```diff
@@ -0,0 +1,94 @@
+{
+    internal class BrokerCredential : InteractiveBrowserCredential
+    {
+        private const string CredentialsSection = "Broker";
```
Do we need this?

scottaddie commented on sdk/identity/Azure.Identity.Broker/tests/DefaultAzureCredentialFactoryTests.cs:
```diff
@@ -57,7 +58,7 @@ public void ValidateDefaultAzureCredentialAZURE_TOKEN_CREDENTIALS_Honored_WithBr
                     Assert.IsTrue(chain.Any(cred => cred is VisualStudioCredential));
                     Assert.IsTrue(chain.Any(cred => cred is AzureDeveloperCliCredential));
                     Assert.IsTrue(chain.Any(cred => cred.GetType() == typeof(VisualStudioCodeCredential)));
```
Nit: While you're in here, update this line of code for consistency.

```suggestion
                    Assert.IsTrue(chain. Any(cred => cred is VisualStudioCodeCredential));
```

christothes commented on sdk/identity/Azure.Identity/src/Credentials/BrokerCredential.cs:
```diff
@@ -0,0 +1,94 @@
+            return null;
+        }
+
+        private static InteractiveBrowserCredentialOptions CreateFallbackOptionsFromCredentialOptions(DevelopmentBrokerOptions credentialOptions)
```
Can we de-dupe this code?

JonathanCrd committed:
Add ISupportsTenantId interface to credential options for tenant ID support

JonathanCrd committed:
Refactor broker options mapping to new helper class

Moved logic for mapping credential options to broker options from BrokerCredential to a new CredentialOptionsMapper class. This improves code organization and reusability by centralizing the mapping logic.

JonathanCrd committed:
remove CredentialsSection from broker

JonathanCrd committed:
Simplify checks

JonathanCrd committed:
Remove duplicated code

JonathanCrd committed:
Adjust tests to new simplified logic

JonathanCrd committed:
Merge branch 'main' into identity-always-add-broker-in-dac

JonathanCrd committed:
Add Broker back after merge

JonathanCrd committed:
Remove unnecessary code for Broker after merge with main

JonathanCrd committed:
Update copilot tests from main with the right amount of dev credentials

christothes commented on sdk/identity/Azure.Identity.Broker/tests/DevelopmentBrokerOptionsTests.cs:
```diff
@@ -20,15 +20,14 @@ public void TryCreateDevelopmentBrokerOptions()
-            {
-                // This is expected, as the broker is not available in the test environment.
+            var cred = factory.CreateBrokerCredential();
+                try
+                {
+                    cred.GetToken(new TokenRequestContext(new[] { "https://management.azure.com/.default" }), default);
+                }
+                catch (CredentialUnavailableException)
+                {
+                    // This is expected, as the broker is not available in the test environment.
```
nit: indenting seems off here.

christothes commented on sdk/identity/Azure.Identity/src/Credentials/CredentialOptionsMapper.cs:
```diff
@@ -0,0 +1,122 @@
+    /// <param name="fileSystem">The file system service to use for reading authentication records.</param>
+    /// <param name="isBrokerEnabled">Output parameter indicating whether broker options are available.</param>
+    /// <returns>The broker options if available, null otherwise.</returns>
+    internal static InteractiveBrowserCredentialOptions TryGetBrokerOptions(out bool isBrokerEnabled, IFileSystemService fileSystem = null)
```
do we need this implementation? The one below seems like it would do the same thing when credentialOptions is null.

christothes commented on sdk/identity/Azure.Identity/src/Credentials/CredentialOptionsMapper.cs:
```diff
@@ -0,0 +1,122 @@
+    /// </summary>
+    /// <param name="credentialOptions"> The credential options to copy properties from.</param>
+    /// <returns> A minimal InteractiveBrowserCredentialOptions instance with properties copied from the provided credential options.</returns>
+    internal static InteractiveBrowserCredentialOptions CreateFallbackOptionsFromCredentialOptions(TokenCredentialOptions credentialOptions)
```
I think we can remove the `CreateFallbackOptions` above if we make `credentialOptions` optional here.

christothes commented on sdk/identity/Azure.Identity/src/Credentials/ISupportsTenantId.cs:
```diff
@@ -0,0 +1,16 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Azure.Identity.Credentials
```
let's keep this in the `Azure.Identity` namespace

JonathanCrd commented on sdk/identity/Azure.Identity/src/DefaultAzureCredentialFactory.cs:
```diff
@@ -343,22 +332,14 @@ public virtual TokenCredential CreateInteractiveBrowserCredential()
         }
 
-        public TokenCredential CreateBrokerCredential(InteractiveBrowserCredentialOptions brokerOptions)
+        internal TokenCredential CreateBrokerCredential()
```
Yes, it's used in this factory class to add the Broker into the chain

christothes replied:
What I meant was all it does is return `new BrokerCredential(options)` - can callers just do that themselves?

JonathanCrd committed:
Applied suggestions and adressed comments

christothes commented on sdk/identity/Azure.Identity/src/Credentials/DeviceCodeCredentialOptions.cs:
```diff
@@ -5,13 +5,14 @@
 using System.Collections.Generic;
 using System.Threading;
 using System.Threading.Tasks;
+using Azure.Identity.Credentials;
```
All these added using statements can be removed.

JonathanCrd committed:
Remove unnecesary using statements

scottaddie commented on sdk/identity/Azure.Identity.Broker/tests/DefaultAzureCredentialFactoryTests.cs:
```diff
@@ -56,8 +57,8 @@ public void ValidateDefaultAzureCredentialAZURE_TOKEN_CREDENTIALS_Honored_WithBr
-                    Assert.IsTrue(chain.Any(cred => cred.GetType() == typeof(VisualStudioCodeCredential)));
-                    Assert.IsTrue(chain.Any(cred => cred.GetType() == typeof(InteractiveBrowserCredential)));
+                    Assert.IsTrue(chain.Any(cred => cred is VisualStudioCodeCredential));
+                    Assert.IsTrue(chain.Any(cred => cred is BrokerCredential));
```
Are there other tests in here that should be updated for BrokerCredential? For example, `ValidateExcludeOptionsHonoredWithAZURE_TOKEN_CREDENTIALS_DevMode`.

christothes replied:
Yes

JonathanCrd committed:
Update tests to use new `BrokerCredential` type

JonathanCrd committed:
Merge branch 'main' into identity-always-add-broker-in-dac

JonathanCrd committed:
Fix tests and build errors after merge

JonathanCrd committed:
Add missing check for ValidateDefaultAzureCredentialAZURE_TOKEN_CREDENTIALS_Honored_WithBroker test

JonathanCrd committed:
Rebase DefaultAzureCredentialFactoryTests.cs and modify tests to expect Broker as part of the chain

JonathanCrd commented on sdk/identity/Azure.Identity/tests/DefaultAzureCredentialFactoryTests.cs:
```diff
@@ -579,7 +545,37 @@ public void ValidateDefaultAzureCredentialAZURE_TOKEN_CREDENTIALS_Honored_WithDa
                 }
+                else if (credSelection == Constants.VisualStudioCredential)
+                {
+                    Assert.IsTrue(chain.Single(cred => cred is VisualStudioCredential) is VisualStudioCredential);
```
I rebased the branch but retained the changes where BrokerCredential was introduced, since it's now a permanent part of the DAC chain when unless excluded.

christothes commented on sdk/identity/Azure.Identity/src/Credentials/InteractiveBrowserCredentialOptions.cs:
```diff
@@ -5,14 +5,15 @@
 using System.Collections.Generic;
 using System.Diagnostics.CodeAnalysis;
 using System.Threading;
+using Azure.Identity.Credentials;
```
We shouldn't need this in any of the options types

JonathanCrd commented on sdk/identity/Azure.Identity/tests/DefaultAzureCredentialFactoryTests.cs:
```diff
@@ -444,13 +449,6 @@ public void ValidateDefaultAzureCredentialAZURE_TOKEN_CREDENTIALS_Honored(string
-                if (credSelection == Constants.BrokerCredential)
-                {
-                    // BrokerCredential is not supported without the Azure.Identity.Broker package.
-                    var ex = Assert.Throws<CredentialUnavailableException>(() => factory.CreateCredentialChain());
```
I can add a new test for this

JonathanCrd committed:
Adressed comments: removed `using Azure.Identity.Credentials` and add new test
